invoiceregistry.go

```diff
@@ -732,54 +659,73 @@ func (i *InvoiceRegistry) process(ctx context.Context, h *registryHtlc) error {
 		return nil
 	}
 
-	state, ok := i.invoices[h.rHash]
-	if !ok {
-		// Invoice is not present in memory. Do a db lookup to see if this
-		// happens to be a previously settled invoice.
-		dbInvoice, htlcs, err := i.cdb.Get(ctx, h.rHash)
-		switch {
-		case err == types.ErrInvoiceNotFound:
-			// If the invoice was not found, return a failure
-			// resolution with an invoice not found result.
-			h.resolve(NewFailResolution(ResultInvoiceNotFound))
+	// Re-derive preimage from payment address.
+	mac := hmac.New(sha256.New, i.cfg.PrivKey[:])
 
-			return nil
+	paymentAddr := mpp.PaymentAddr()
+	mac.Write(paymentAddr[:])
+	paymentPreimage, err := lntypes.MakePreimage(mac.Sum(nil))
+	if err != nil {
+		return err
+	}
 
-		case err != nil:
-			return err
-		}
+	// If the preimage doesn't match the payment hash, fail this htlc. Someone
+	// must have tampered with our payment parameters.
+	if paymentPreimage.Hash() != h.rHash {
+		logger.Debugw("Hash mismatch")
 
-		// Fail htlcs paying to unpayable invoices (expired or cancelled).
-		if dbInvoice.State != persistence.InvoiceStateSettleRequested &&
-			dbInvoice.State != persistence.InvoiceStateSettled {
+		h.resolve(NewFailResolution(ResultInvoiceNotFound))
 
-			h.resolve(NewFailResolution(ResultInvoiceNotFound))
+		return nil
+	}
 
-			return nil
-		}
+	// Extract payment parameters.
+	expiry := time.Unix(int64(byteOrder.Uint64(paymentAddr[16:24])), 0)
+	amtMsat := byteOrder.Uint64(paymentAddr[24:32])
 
-		i.logger.Debugw("Loaded settled invoice from db", "hash", h.rHash)
+	logger = logger.With(
+		"amtMsat", amtMsat, "expiry", expiry,
+	)
 
-		// Handle replays to a settled invoice.
-		if len(htlcs) == 0 {
-			return errors.New("unexpected unsettled invoice")
-		}
+	// Check expiry.
+	if expiry.Before(time.Now()) {
+		logger.Infow("Stateless invoice payment to expired invoice")
 
-		// If this htlc was used for settling the invoice,
-		// resolve to settled again.
-		if _, ok := htlcs[h.circuitKey]; ok {
-			h.resolve(NewSettleResolution(
-				dbInvoice.PaymentPreimage,
-				ResultReplayToSettled,
-			))
+		h.resolve(NewFailResolution(ResultInvoiceExpired))
 
-			return nil
+		return nil
+	}
+
+	logger.Infow("Stateless invoice payment received")
+
+	// Look up this invoice in memory. If it is present, we have already
+	// received other shards of the payment.
+	state, ok := i.invoices[h.rHash]
+	if !ok {
+		state = &invoiceState{
+			invoice: &types.InvoiceCreationData{
+				PaymentPreimage: paymentPreimage,
+				Value:           lnwire.MilliSatoshi(amtMsat),
+				PaymentAddr:     paymentAddr,
+			},
+			acceptedHtlcs: make(map[types.CircuitKey]*InvoiceHTLC),
+			expiry:        expiry,
 		}
 
-		// Otherwise fail the htlc.
-		h.resolve(NewFailResolution(ResultInvoiceNotOpen))
+		i.invoices[h.rHash] = state
+	} else {
+		// Sanity check that the total amount and expiry time are identical.
+		if amtMsat != uint64(state.invoice.Value) ||
```

Comment:
Is this even possible?

Reply:
Normally not. Unless a bug has occurred or our private key was compromised and used to fabricate invalid payment addresses.

Reply:
If any of those happen, shouldn't we `.Fatal`? IMO it's fair to stop the execution in any of the cases you describe (bug or pk compromised)

Reply:
Yes I get that. But wondering what the best communication path is for that. I could introduce prometheus metrics in this repo for example, if that is considered a good option.

Reply:
As mentioned offline, I am suggesting to add monitoring in a follow up pr. Think a bit deeper about it and also add other metrics that are relevant.

Reply:
https://github.com/bottlepay/lnmux/pull/42